### PR TITLE
feat(target_chains/fuel): add governance contract

### PR DESCRIPTION
**Pull request by cctdaniel**
`UpgradeContract` is not implemented yet because Fuel have not released an official upgrade standard, the rest of the implementation are referenced from the current implementations of other chains e.g. evm/solana/near

### CODE REVIEW

File: target_chains/fuel/contracts/pyth-contract/src/data_structures/data_source.sw [hunk 1]
```diff
@@ -19,7 +19,7 @@ impl DataSource {
     }
 
     #[storage(read)]
-    pub fn is_valid(
+    pub fn is_valid_data_source(
```
Comment (cctdaniel) [inline target_chains/fuel/contracts/pyth-contract/src/data_structures/data_source.sw:22]: drive-by: change name to be more descriptive

File: target_chains/fuel/contracts/tests/functions/pyth_governance/execute_governance_instruction.rs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+            default_data_sources(),
+            GOVERNANCE_DATA_SOURCE,
+            WORMHOLE_GOVERNANCE_DATA_SOURCE,
+            DEFAULT_SINGLE_UPDATE_FEE,
+            DEFAULT_VALID_TIME_PERIOD,
+            bits256_guardians,
+            0,
+            DUMMY_CHAIN_ID,
+        )
+        .await;
```
Comment (ali-bahjati) [inline target_chains/fuel/contracts/tests/functions/pyth_governance/execute_governance_instruction.rs:61]: Maybe move this setup code to another function and then have separate tests for each governance action?

**Review by ali-bahjati — APPROVED** [review]

File: target_chains/fuel/contracts/pyth-contract/src/main.sw [hunk 3]
```diff
@@ -615,3 +703,219 @@ fn submit_new_guardian_set(encoded_vm: Bytes) {
...
+}
+
+#[storage(read, write)]
+fn set_data_sources(payload: SetDataSourcesPayload) {
+    let old_data_sources = storage.valid_data_sources.load_vec();
+
+    let mut i = 0;
+    while i < old_data_sources.len {
+        let data_source = old_data_sources.get(i).unwrap();
+        storage.is_valid_data_source.insert(data_source, false);
```
[the remark this answers is not in the document]
Comment (cctdaniel) [reply target_chains/fuel/contracts/pyth-contract/src/main.sw:747]: it seems like in Sway, `StorageMap` does not provide a direct method to retrieve all keys like some other languages' map or dictionary structures might which is why I think we need this additional variable

File: target_chains/fuel/contracts/pyth-contract/src/main.sw [hunk 4]
```diff
@@ -409,13 +435,67 @@ fn valid_time_period() -> u64 {
...
+        wormhole_governance_data_source: DataSource,
         single_update_fee: u64,
         valid_time_period_seconds: u64,
-        wormhole_guardian_set_upgrade: Bytes,
+        wormhole_guardian_set_addresses: Vec<b256>,
+        wormhole_guardian_set_index: u32,
+        chain_id: u16,
     ) {
         initialize_ownership(DEPLOYER);
         only_owner();
```
[the remark this answers is not in the document]
Comment (cctdaniel) [reply target_chains/fuel/contracts/pyth-contract/src/main.sw:491]: updated the code with comments to explain them but writing it here again
```
        // This function sets up the necessary state in the contract to recognize the DEPLOYER as the owner. https://github.com/FuelLabs/sway-libs/blob/8045a19e3297599750abdf6300c11e9927a29d40/libs/src/ownership.sw#L127-L138
        initialize_ownership(DEPLOYER);
        // This function checks if the DEPLOYER is the owner of the contract. https://github.com/FuelLabs/sway-libs/blob/8045a19e3297599750abdf6300c11e9927a29d40/libs/src/ownership.sw#L59-L65
        only_owner();
```
